test: remove assert.doesNotThrow()

BridgeAR opened the pull request:
There is actually no reason to use `assert.doesNotThrow()` in the tests. If a test threw, just let the error bubble up right away instead of first catching it and then rethrowing it. That way we get the best stack trace and normally also the best possible error message.

Someone might argue that there are rare cases (I guess about 10) where it would actually make sense to keep the current behavior. The reason would be that a loop is used to execute something and a generated error message would be thrown that is individual per entry. Otherwise it might be hard to know what entry failed.

Nevertheless, I think a better way to deal with those would be to add a console.log. And we can add those as soon as something fails.

I can not think of any other reason to keep `assert.doesNotThrow()`. I personally feel we might want to deprecate that function actually.

I kept the calls in the assertion test file because I work on that right now in a different PR and those mainly test `assert.doesNotThrow()` on its own.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
test

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/13051/

bnoordhuis commented on test/addons/symlinked-module/test.js:
```diff
@@ -30,5 +30,5 @@ const sub = require('./submodule');
   assert.notStrictEqual(mod, null);
   assert.strictEqual(mod.hello(), 'world');
-  assert.doesNotThrow(() => sub.test(i));
+  sub.test(i);
```
In a couple of places (like here) the `doesNotThrow()` is really a comment in disguise.

Someone reading the new code might be confused and think the `sub.test(i)` is a bug or dead code because the return value isn't checked, not realizing it's there to check a side effect (throwing an exception.)

I would suggest rewriting lines like these to `sub.test(i);  // Should not throw.`.  Apart from that getting rid of `doesNotThrow()` is (IMO) a good move.

BridgeAR replied:
I went throw the code and added comments were I felt it made sense. I also some times improved a few tests along the way and added punctuation and caps and things like that.

BridgeAR commented:
I went through the code and added comments were I felt it made sense that it should not throw. Some times I also improved a few tests along the way and added punctuation and caps and things like that. I also removed one that was outdated.

BridgeAR commented:
Rebased due to conflicts.

Light-CI https://ci.nodejs.org/job/node-test-commit-light/252/

BridgeAR commented:
I added two more commits: the first one cleans up the rest of the `doesNotThrow` statements that should not be there (I did not do most of those before because of a otherwise conflicting PR that landed in the meanwhile).

The second commit adds a eslint rule to prohibit the usage of `assert.doesNotThrow`. I could of course also remove that commit again and open a separate PR for it after this lands. I just thought it can probably also be included here. Is everyone fine with that?

BridgeAR commented:
Light-CI https://ci.nodejs.org/job/node-test-commit-light/257/

BridgeAR commented:
Rebased due to conflicts.

BridgeAR commented:
Light-CI https://ci.nodejs.org/job/node-test-commit-light/258/

BridgeAR commented:
New Light-CI https://ci.nodejs.org/job/node-test-commit-light/264/

BridgeAR commented:
Landed in 4d3c3f039af08b954fbbba1e9a50979ffc98592b...15bb8437fd0a0eb07845ccf746c9ecb47fcfcb25

MylesBorins commented:
Should this be backported to `v9.x-staging`? If yes please follow [the guide](https://github.com/nodejs/node/blob/master/doc/guides/backporting-to-release-lines.md) and raise a backport PR, if not let me know or add the `dont-land-on` label.

MylesBorins commented:
The backport should also include https://github.com/nodejs/node/pull/18854

BridgeAR commented:
Backported in #19244

@MylesBorins I am not certain in what way this correlates to the mentioned PR?

MylesBorins commented:
Backport requested in https://github.com/nodejs/node/pull/19244
